### ISSUE TITLE
security: Propose alternate TLS verification modes for postgres clients

### ISSUE THREAD

**Issue by bdarnell** [cockroachdb/cockroach#32932]
Dealing with root CA certs in client applications (that use password-based authentication) is cumbersome. Instead of a self-contained connection string, the application deployment must either install a CA cert to the system certificate store (not recommended) or place it on disk somewhere and then reference that path from the connection string (which may mean constructing it dynamically instead of using static configuration). 

This process could be simplified if there were some way to encode the CA certificate or some equivalent verification method into the connection string itself. We could propose and implement a way to do this in `libpq` and its counterparts in other languages (note that the proposed extensions would not be cockroach-specific; any hosted postgresql offering could use them as well. Amazon RDS and google cloud sql do not use TLS by default, and I suspect that this is part of the reason).

The simplest way to do this would be to add a new connection string parameter `sslrootcertdata`, which contains the base64-encoded contents of the `ca.crt` file. This would produce extremely large connection strings, but it's just a single value to copy and paste so it might be easier in practice than dealing with a separate file.

For a more ambitious extension which would produce more reasonably-sized connection strings, we could introduce a new `sslmode` verification option based on public key pinning. [HPKP](https://tools.ietf.org/html/rfc7469) failed on the web because of mismatches between PKP and the CA-based security model, but for a managed database offering (with single-use CAs and secure channels for distributing the public key pins) the key pinning model may actually be a better fit than the CA model.

Briefly, the idea is that setting `sslmode=verify-pkp` (instead of `verify-full` or `verify-ca`) would cause the client to verify the certificate according to a new `sslpkp` parameter which contains a string similar to the HPKP `Public-Key-Pins` header (I'm just copying that syntax for now as a strawman). The query string would look something like this:

    postgresql://user@db.example.com:26257/db?sslmode=verify-pkp&sslpkp=pin-sha256="d6qzRu9zOECb90Uez27xWltNsj0e1Md7GkYYkVoZWmM=",pin-sha256="LPJNul+wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQ="

(there would only be a single pin in many cases, but when a certificate nears expiration and is rotated, two or more pins would be needed)

Doing anything like this would require buy-in from the postgres community; I'm posting this on the cockroach issue tracker first to allow us to discuss and decide if this is something we want to push for. The benefits from this change would not be felt for some time since updated versions of client libraries would need to propagate through the ecosystem (it could be years before versions of `libpq` supporting this extension become widespread enough that we could confidently recommend this pattern over the `ca.crt` file), and it could result in fragmentation if some languages' client libraries are slower to update.

**Comment by knz** [comment]
Libraries have length limits on URL strings. This issue would be less painful if the pg drivers were able to use other algorithms than RSA -- for example EC keys are much smaller.

**Comment by bdarnell** [comment]
Note that to replace the CA certificate, we just need a hash, not a full key. This results in a full URL of a few hundred bytes, well below the `~2KB` that is the lowest commonly-cited limit for URLs on the web.

**Comment by knz** [comment]
I think this proposal becomes much more tractable if we start to develop/recommend custom crdb drivers?

@rafiss do you know if we're making such a plan already?

**Comment by bdarnell** [comment]
Yes, if we make our own drivers we can do this or many other things (for example, we could recognize cockroach cloud URLs and pull the per-cluster CA certificate automatically, bootstrapping from the web root CAs)

**Comment by rafiss** [comment]
We won't be making a custom driver during the next few quarters, but it is generally something we want to work towards. A few of our users are keen to have one for one reason or another. An issue like this could be a good impetus for making one, as we might be able to start off by forking some other Postgres driver, and adding a feature like this in.

**Comment by knz** [comment]
cc @jtsiros for triage and tracking
